nomad/state/state_store.go

```diff
@@ -7348,6 +7363,30 @@ func (s *StateStore) ScalingPoliciesByIDPrefix(ws memdb.WatchSet, namespace stri
 	return iter, nil
 }
 
+func isReschedulable(a *structs.Allocation) bool {
+	if a.ReschedulePolicy() == nil || a.Job.Type != structs.JobTypeService {
+		return false
+	}
+
+	reschedulePolicy := a.ReschedulePolicy()
+	availableAttempts := reschedulePolicy.Attempts
+	interval := reschedulePolicy.Interval
+	attempted := 0
+	currTime := time.Now()
+
+	// Loop over reschedule tracker to find attempts within the restart policy's interval
+	if a.RescheduleTracker != nil && availableAttempts > 0 && interval > 0 {
+		for j := len(a.RescheduleTracker.Events) - 1; j >= 0; j-- {
+			lastAttempt := a.RescheduleTracker.Events[j].RescheduleTime
+			timeDiff := currTime.UTC().UnixNano() - lastAttempt
+			if timeDiff < interval.Nanoseconds() {
+				attempted += 1
+			}
+		}
+	}
+	return attempted < availableAttempts
+}
```

Comment:
There's a method `(*RescheduleTracker).RescheduleEligible` we should be using here so that we're sharing identical logic between the state store and scheduler.